salt/transport/ipc.py

```diff
@@ -220,12 +263,186 @@ def _connect(self):
+                    msg = salt.transport.frame.frame_msg(
+                        message_future._msg,
+                        raw_body=True
+                    )
+                    yield self.stream.write(msg)
+                    self.remove_message_timeout(message_future)
+                    message_future.set_result(True)
+                else:
+                    log.critical('Unknown message kind {0}'.format(message_future._msg_kind))
+
+            # if the connection is dead, lets fail this send, and make sure we
+            # attempt to reconnect
+            except tornado.iostream.StreamClosedError as e:
+                self.inflight_messages.pop(message_id).set_exception(Exception())
+                self.remove_message_timeout(message_future)
+                # if the last connect finished, then we need to make a new one
+                self.disconnect()
+
+    def timeout_message(self, message_future):
+        '''
+        timeout a given message_id
+        '''
+        del self.timeout_map[message_future]
+        if message_future in self.send_queue:
+            log.trace('Timing out future from send queue')
+            self.send_queue.remove(message_future)
+        else:
+            log.trace('Timing out inflight future from send queue')
+            self.inflight_messages.pop(message_future._msg_id).set_exception(SaltReqTimeoutError('Message timed out'))
+
+    def remove_message_timeout(self, message_future):
+        if message_future not in self.timeout_map:
+            return
+        timeout = self.timeout_map.pop(message_future)
+        self.io_loop.remove_timeout(timeout)
+
+    def send(self, msg, timeout=None, callback=None):
+        '''
+        Send given message, and return a future with the result of the message
+        '''
+        future = tornado.concurrent.Future()
+        future._msg = msg  # TODO: create a future subclass?
+        future._msg_kind = 'send'
+        if callback is not None:
+            def handle_future(future):
+                response = future.result()
+                self.io_loop.add_callback(callback, response)
+            future.add_done_callback(handle_future)
+
+        if timeout is not None:
+            send_timeout = self.io_loop.call_later(timeout, self.timeout_message, future)
+            self.timeout_map[future] = send_timeout
+
+        # if we don't have a send queue, we need to spawn the callback to do the sending
+        if len(self.send_queue) == 0:
+            self.io_loop.spawn_callback(self._handle_outgoing)
+        self.send_queue.append(future)
+        return future
+
+    def publish(self, msg, timeout=None, callback=None):
+        '''
+        Publish a message (send without expecting a response. Return a future
+        which will complete when the message has been sent
+        '''
+        future = tornado.concurrent.Future()
+        future._msg = msg  # TODO: create a future subclass?
```

Comment:
Well, right now I'm setting some attributes in the future to store some data. Making something like MessageFuture(Future) which has those more defined might be more correct ;)